backend/danswer/indexing/chunker.py

```diff
@@ -122,192 +72,247 @@ def _get_metadata_suffix_for_document_index(
                     title_prefix=title_prefix,
                     metadata_suffix_semantic=metadata_suffix_semantic,
                     metadata_suffix_keyword=metadata_suffix_keyword,
-                    mini_chunk_texts=mini_chunk_splitter.split_text(chunk_text)
-                    if enable_mini_chunk and chunk_text.strip()
-                    else None,
+                    mini_chunk_texts=self._get_mini_chunk_texts(chunk_text),
                 )
             )
-            link_offsets = {0: section_link_text}
-            chunk_text = section_text
 
-    # Once we hit the end, if we're still in the process of building a chunk, add what we have. If there is only whitespace left
-    # then don't include it. If there are no chunks at all from the doc, we can just create a single chunk with the title.
-    if chunk_text.strip() or not chunks:
-        chunks.append(
-            DocAwareChunk(
-                source_document=document,
-                chunk_id=len(chunks),
-                blurb=extract_blurb(chunk_text, blurb_splitter),
-                content=chunk_text,
-                source_links=link_offsets,
-                section_continuation=False,
-                title_prefix=title_prefix,
-                metadata_suffix_semantic=metadata_suffix_semantic,
-                metadata_suffix_keyword=metadata_suffix_keyword,
-                mini_chunk_texts=mini_chunk_splitter.split_text(chunk_text)
-                if enable_mini_chunk and chunk_text.strip()
-                else None,
+        # If the chunk does not have any useable content, it will not be indexed
+        return chunks
+
+    def chunk(self, document: Document) -> list[DocAwareChunk]:
+        # Specifically for reproducing an issue with gmail
+        if document.source == DocumentSource.GMAIL:
+            logger.debug(f"Chunking {document.semantic_identifier}")
+
+        title = self._extract_blurb(document.get_title_for_document_index() or "")
+        title_prefix = title + RETURN_SEPARATOR if title else ""
+        title_tokens = len(self.tokenizer.tokenize(title_prefix))
+
+        metadata_suffix_semantic = ""
+        metadata_suffix_keyword = ""
+        metadata_tokens = 0
+        if self.include_metadata:
+            (
+                metadata_suffix_semantic,
+                metadata_suffix_keyword,
+            ) = _get_metadata_suffix_for_document_index(
+                document.metadata, include_separator=True
             )
+            metadata_tokens = len(self.tokenizer.tokenize(metadata_suffix_semantic))
+
+        if metadata_tokens >= self.chunk_token_limit * MAX_METADATA_PERCENTAGE:
+            # Note: we can keep the keyword suffix even if the semantic suffix is too long to fit in the model
+            # context, there is no limit for the keyword component
+            metadata_suffix_semantic = ""
+            metadata_tokens = 0
+
+        content_token_limit = self.chunk_token_limit - title_tokens - metadata_tokens
```

Comment:
Where does content_token_limit get used?

Reply:
Good catch, i pass it into the chunk logic instead of using the class variable